pydantic/validate_call_decorator.py

```diff
@@ -50,11 +50,7 @@ def validate_call(
     local_ns = _typing_extra.parent_frame_namespace()
 
     def validate(function: AnyCallableT) -> AnyCallableT:
-        if isinstance(function, (classmethod, staticmethod)):
-            name = type(function).__name__
-            raise TypeError(f'The `@{name}` decorator should be applied after `@validate_call` (put `@{name}` on top)')
-
-        validate_call_wrapper = _validate_call.ValidateCallWrapper(function, config, validate_return, local_ns)
+        validate_call_wrapper = _validate_call.wrap_validate_call(function, config, validate_return, local_ns)
```

Comment:
I wonder if it makes sense to leave the function type checking logic in this file, just from a readability perspective.

Baiscally, check if `validate_call` supports wrapping said function, then define the validate function (see below)

Reply:
Side note - what's the case where `func` is None?

Reply:
Should we further abstract away some of this:

```py
if inspect.iscoroutinefunction(function):
    @functools.wraps(function)
    async def wrapper_function(*args, **kwargs):  # type: ignore
        return await validate_call_wrapper(*args, **kwargs)
else:
    @functools.wraps(function)
    def wrapper_function(*args, **kwargs):
        return validate_call_wrapper(*args, **kwargs)
wrapper_function.raw_function = function  # type: ignore
return wrapper_function  # type: ignore
```

into the `_validate_call.py` file?

Reply:
I think the type-checking should be inside `validate` since `func` could be `None` anyways. Maybe we could rename `validate` if it is for the sake of readability.

Reply:
> Side note - what's the case where func is None?

Do you know what the case is for this?

Reply:
This is for decorating without any args:
```py
@validate_call
def f(): ...
```

So in this we do not yet have the actual function inside `validate_call`, and we have to return a function which serves as a decorator.